### PR TITLE
Doc generation

### PR DESCRIPTION
**Pull request by fragsalat**
Pull request for issue: #81 #62 

Description:
This pr contains fixed generation of API documentation and integration into the circle ci build process.

### CODE REVIEW

File: build/tasks/docs.js [hunk 1]
```diff
@@ -6,45 +6,57 @@ const glob = require('glob');
 const del = require('del');
 const vinylPaths = require('vinyl-paths');
 const docjs = require('documentation');
+const REGEX_FILE_NAME = /((\w)+(-)*(\w))+(?=.js)/g;
```
Comment (faboweb) [inline build/tasks/docs.js]: just a comment: I like REGEX too, but they are damn hard to read.

File: build/tasks/docs.js [hunk 2]
```diff
@@ -6,45 +6,57 @@ const glob = require('glob');
...
-    const folder = `${paths.docsOutput}${mdDoc.folder}`;
-    const filePath = mdDoc.file !== 'index' ? `${paths.docsOutput}${mdDoc.folder}/${mdDoc.file}.md` : `${paths.docsOutput}${mdDoc.file}.md`;
-    if (!fs.existsSync(folder)) {
-      fs.mkdirSync(folder);
-    }
-    return fs.writeFile(filePath, mdDoc.markdown);
-  });
-  return Promise.all(writeMdFiles);
+gulp.task('build-jsdoc-to-md', done => {
+  Promise.resolve(glob.sync(paths.source))
```
Comment (faboweb) [inline build/tasks/docs.js]: nice chain 👍

File: build/tasks/docs.js [hunk 3]
```diff
@@ -6,45 +6,57 @@ const glob = require('glob');
...
+      groups[doc.folder] = groups[doc.folder] || [];
+      groups[doc.folder].push(doc);
+      return groups;
+    }, {}))
+    // Create markdown files for each folder
+    .then(groups => Object.keys(groups).forEach(folderName => {
+      const group = groups[folderName];
+      const fileName = `${paths.apiDoc}/${folderName}.md`;
+      const normalize = doc => doc.markdown.replace(/<!--.*?-->\n*/g, '');
+      const content = `# ${folderName}\n${group.map(normalize).join('')}`;
```
Comment (faboweb) [inline build/tasks/docs.js]: why the hash?

File: build/tasks/lint.js [hunk 4]
```diff
@@ -1,9 +1,25 @@
 var gulp = require('gulp');
 var paths = require('../paths');
 var eslint = require('gulp-eslint');
+const glob = require('glob');
+const docjs = require('documentation');
+const REGEX_FILE_NAME = /((\w)+(-)*(\w))+(?=.js)/g;
```
Comment (faboweb) [inline build/tasks/lint.js]: maybe give an acceptable example

File: build/tasks/lint.js [hunk 5]
```diff
@@ -1,9 +1,25 @@
...
 
 // runs eslint on all .js files
 gulp.task('lint', function() {
+  Promise.resolve(glob.sync(paths.source))
+    // Filter file names to match tag name pattern
+    .then(fileNames => fileNames.filter(fileName => {
+      const name = fileName.match(REGEX_FILE_NAME)[0];
+      return name !== 'imports';
```
Comment (faboweb) [inline build/tasks/lint.js]: what are the "imports" files and why are they ignored?

File: build/tasks/lint.js [hunk 6]
```diff
@@ -1,9 +1,25 @@
...
+    }))
+    // Generate markdown for each file
+    .then(fileNames => fileNames.forEach(fileName =>
+      docjs.lint(fileName, {}).then(lintOutput => {
```
Comment (faboweb) [inline build/tasks/lint.js]: is this generating markdown? it says lint?!

File: build/tasks/docs.js (hunk 1 above)
Comment (fragsalat) [inline build/tasks/docs.js]: I just moved them up to a global constant. This regex seem to match a file name pattern like this ws-something-somehting.js
Maybe it could be done simpler but there was no need to simplify it right now in my eyes^^

File: doc/components/dropdown.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
+# Dropdown
```
Comment (faboweb) [inline doc/components/dropdown.md]: I would have put new docs in a separate PR

File: src/ws-dropdown/ws-dropdown.js [hunk 8]
```diff
@@ -229,7 +229,7 @@ export class WSDropdown extends Component {
       itemsToWrap = items ? [items] : [];
     }
     return itemsToWrap.map(item => {
-      const enriched = typeof item === 'object' ? item : {label: item};
+      const enriched = typeof item === 'object' ? item : {label: item, value: item};
```
Comment (faboweb) [inline src/ws-dropdown/ws-dropdown.js]: the same with small fixes :)

File: build/tasks/docs.js (hunk 1 above)
Comment (fokusferit) [inline build/tasks/docs.js]: I checked it again, as it was my regex (feels over-engineered :D).

Could you just try out this one: ` /((-)*(\w))+(?=.js)/g`instead of the one above. What it basically does is, it 'captures' all words of a filename until ".js". It also checks for a beginning dash of the name so it can capture the whole filename. If we would just use "\w (word)" it would only catch the first ws and stop.

We could write some code but this would be:

- Split filename `split(-)`;
- Split the last Element of this Array again `split(.)` as array would be `['ws', 'input', 'edit.js']`
- create a new Array combining both
- `.join(-)` to get "ws-input-edit".

But I thought a one line regex makes sense here instead of doing this slicing, splitting thing ^^.

File: build/tasks/docs.js (hunk 3 above)
Comment (fokusferit) [inline build/tasks/docs.js]: I think this is related to how the API documentation is later included in the styleguide.
I remember that @fragsalat told me that his markdown-interpreter for the styleguide needs this "#" which is basically markdown headline (h1) :D

File: build/tasks/docs.js [hunk 9]
```diff
@@ -6,45 +6,57 @@ const glob = require('glob');
...
+    .then(docs => docs.reduce((groups, doc) => {
+      groups[doc.folder] = groups[doc.folder] || [];
+      groups[doc.folder].push(doc);
+      return groups;
+    }, {}))
+    // Create markdown files for each folder
+    .then(groups => Object.keys(groups).forEach(folderName => {
+      const group = groups[folderName];
+      const fileName = `${paths.apiDoc}/${folderName}.md`;
+      const normalize = doc => doc.markdown.replace(/<!--.*?-->\n*/g, '');
```
Comment (fokusferit) [inline build/tasks/docs.js]: +1 for removing docjs comment ^^

File: build/tasks/lint.js (hunk 5 above)
Comment (fokusferit) [inline build/tasks/lint.js]: I would say there is no need to ignore it. imports.js is where we abstract the our React / Preact imports from node_modules.

File: build/tasks/lint.js (hunk 6 above)
Comment (fokusferit) [inline build/tasks/lint.js]: Agreed, I wouldn't use docjs.lint for linting our files even if they use eslint under the hood.

File: src/index.js [hunk 10]
```diff
@@ -5,13 +5,11 @@
  * Here we should add more descriptions
  * Below are all components referenced
  *
- * - See {@link ws-header/ws-header.md|WSHeader}
- * - See {@link ws-date-picker/ws-date-picker.md|WSDatePicker}
- * - See {@link ws-dropdown/ws-dropdown.md|WSDropdown}
- * - See {@link ws-inline-edit/ws-inline-edit.md|WSInlineEdit}
- * - See {@link ws-notification/ws-notification.md|WSNotification}
- * - See {@link ws-week-picker/ws-week-picker.md|WSWeekPicker}
- * - See {@link ws-tiles-chart/ws-tiles-chart.md|WSTilesChart}
```
Comment (fokusferit) [inline src/index.js]: Why did you remove them ?

File: build/tasks/docs.js (hunk 3 above)
Comment (faboweb) [inline build/tasks/docs.js]: we could add this as a comment

Comment (fragsalat) [inline build/tasks/docs.js]: The hash is markdown h1 tag^^

File: build/tasks/lint.js (hunk 5 above)
Comment (fragsalat) [inline build/tasks/lint.js]: I will ignore not only imports but also index.js because the API doc for this is not needed in my eyes.

File: build/tasks/lint.js (hunk 6 above)
Comment (fragsalat) [inline build/tasks/lint.js]: Oh this is a relict and will be removed soon

File: src/ws-dropdown/ws-dropdown.js (hunk 8 above)
Comment (fragsalat) [inline src/ws-dropdown/ws-dropdown.js]: But the fixes appeared during writing the docs^^

File: build/tasks/docs.js (hunk 9 above)
Comment (fragsalat) [inline build/tasks/docs.js]: It is necessary. Otherwise there would be sooo much white space

File: src/index.js (hunk 10 above)
Comment (fragsalat) [inline src/index.js]: Because it would generate an empty page on the styleguide which is not needed. We need the API docs for the classes and not for a simple index file

File: build/tasks/lint.js (hunk 6 above)
Comment (fragsalat) [inline build/tasks/lint.js]: Wrong it is not a relict. This lint's the code for invalid annotations. +1 for moving it into a separate task

File: src/index.js (hunk 10 above)
Comment (fokusferit) [inline src/index.js]: ok, never mind. I think the whole Markdown Index.md was to have a easy to use API documentation hosted on Github. As we move to show them in our styleguide anyway maybe this index.md is unnecessary anymore.
